### ISSUE TITLE
Applying transforms to a batch of images

### ISSUE THREAD

**Issue by sayakpaul** [albumentations-team/albumentations#881]
Hi.

I am aware of [this guide](https://albumentations.ai/docs/examples/example_multi_target/) that shows how to apply transforms to multiple images, masks, etc. But I am not sure how I could extend this to use inside a more generic pipeline with `tf.data`. 

Consider the [existing guide](https://albumentations.ai/docs/examples/tensorflow-example/) that shows how to use `albumentations` with TensorFlow. It applies the augmentation transforms to single examples and for a larger dataset, it can quickly become very expensive. Rather, applying the pipeline on batches might be more efficient as stated [here](https://www.tensorflow.org/guide/data_performance). 

How can we achieve this i.e. how can we modify the augmentation pipeline, in this case, to allow a batch of images to be processed?

**Comment by Dipet** [comment]
Library applies transformations on CPU, we may achieve better performance with batches, but not so much than  if we will use GPU for processing. I think we will see better performance just only if we will use small images.
For this reason, the task of supports batches is of very low priority.

**Comment by sayakpaul** [comment]
I see. But having it supporting batches might beneficial for the `tf.data` pipeline since `tf.numpy_function` is already very expensive.

**Comment by ShairozS** [comment]
I agree this would be nice, if not for performance than just for ease of use. The best case scenario would be to be able to use transforms directly against a batched tensor in Pytorch or TF, but even against a batched np array would make things cleaner. Here's what I'm using right now:


```
import albumentations as A
import numpy as np

def apply_along_batch(func, arr):
    '''
    Apply a callable over a batch of 2+D numpy arrays
    
    args:
        func (callable): A function that takes in a 2+D numpy array and
        returns a numpy array of the same dimensions
        
        arr (numpy.ndarray): A numpy array of the shape (BATCH, H, W, C, ...)
        
    returns:
        (numpy.ndarray): A numpy array with func applied along batch axis (BATCH, H, W, C, ...)
    '''
    res = map(func, arr)
    #res = map(lambda x: np.array(x), res)
    res = list(res)
    return(np.array(res))


arr = np.random.random((128, 64, 64, 3))

transform = A.RandomCrop(width=32, height=32)


res = apply_along_batch(func = lambda x: transform(image=x)['image'], 
                  arr = np.random.random((12, 64,64,3)))
```

**Comment by sayakpaul** [comment]
Thanks much for sharing. How it is performance-wise currently?

**Comment by GuillaumeLeclerc** [comment]
GPU augmentations are very useful in nodes having many GPUs (8-16). In environments with many jobs running it is not acceptable to use 20+ cores for a single GPU job as machines run out of available cores before they run out of GPUs. Having (most of the time) very expensive GPUs idle is a massive waste of resources.
